enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,782 @@
+  // +kubebuilder:validation:Enum=machine.openshift.io;cluster.x-k8s.io
+  // +optional
+  AuthoritativeAPI string `json:"authoritativeAPI,omitempty"`
+
+  // synchronizedGeneration is the generation of the authoritative resource that the non-authoritative resource is synchronised with.
+  // This field is set when the authoritative resource is updated and the sync controller has updated the non-authoritative resource to match.
+  // +kubebuilder:validation:Minimum=0
+  // +optional
+  SynchronizedGeneration int64 `json:"synchronizedGeneration,omitempty"`
+}
+```
+
+These fields will be added to the following resources:
+* Machine
+* MachineSet
+* MachineHealthCheck
+* ControlPlaneMachineSet[^5]
+
+[^5]: The ControlPlaneMachineSet is a Machine API resource that is not present in Cluster API.
+We will extend the `template` field of the ControlPlaneMachineSet to support Cluster API in a separate enhancement.
+These fields will be added when that enhancement is implemented.
+
+### Implementation Details/Notes/Constraints
+
+#### Synchronisation of resources
+
+To ensure that resources are synchronised, we will implement a new controller `machine-api-sync-controller`.
+This controller will be responsible for translation of Machine API resources to Cluster API resources, and vice-versa.
+
+This controller will be responsible for creating Cluster API versions of Machine API resources, when they do not exist.
+It will observe the authoritative `spec` field described [above][authoritative-api] to synchronise between the two resources,
+overwriting anything in the non-authoritative resource that does not match with the authoritative resource.
+
+When errors occur during synchronisation, the controller will write conditions to the Machine API resource to report the failure.
+
+On successful synchronisation, the controller will update the Machine API resource's `status` to reflect the current synchronised generation of the authoritative resource,
+and set the `status.authoritativeAPI` field to the value of the authoritative resource.
+This will allow users to keep track of, and ensure that the latest changes to the authoritative resource have been synchronised.
+
+[authoritative-api]: (#authority-of-resources-and-controller-reconciliations)
+
+##### MachineSet Synchronisation
+
+MachineSets in Machine API map to a MachineSet and InfrastructureTemplate in Cluster API.
+The sync controller will create the Cluster API MachineSet with the same name as the Machine API MachineSet to ensure a 1:1 mapping between these resources.
+The InfrastructureTemplate will be named based on the MachineSet name and a hash of the content within it [^6].
+InfrastructureTemplates are immutable, but MachineSet’s in Machine API are not, should the content of the MachineSet change,
+a new InfrastructureTemplate will be created and the old template will be removed when no longer required.
```

Comment:
Side note, there are facilities upstream CAPI to do rollouts of templates, we should reuse those during implementation